### ISSUE TITLE
A way to retrieve the elapsed time of a job

### ISSUE THREAD

**Issue by Zughy** [minetest/minetest#15047]
### Problem

I'd like to know how much time has passed since an `after` has been run, e.g.
```lua
local timer = minetest.after(3, function() stuff end)
```
where, checking `timer` after 1 sec, will return `1`

### Solutions

I don't know, it could be an internal value of `timer` (in my example)

### Alternatives

---

**Comment by sfan5** [comment]
In-game or real time?

**Comment by Zughy** [comment]
What's the main difference? I'd say the equivalent of launching get_us_time() when the after starts

**Comment by Zughy** [comment]
Usecase: after shooting, my weapons have a recovery time, where no actions but reloading can be performed and players remain slowed down until it ends. An after declares when such recovery state ends (e.g. 1s)

If I start reloading, the weapon state gets overridden and the recovery after interrupted. However, I might cancel the reload by switching weapons, and if that happens within the time frame of the recovery state I've just interrupted (e.g. after 0.5s, with the frame being 1s), I want the weapon state to finish the 0.5 recovery time that was missing, so that players won't go faster 0.5 earlier than they're supposed to. In order to do that, I need to know when the recovery after had been interrupted

**Comment by appgurueu** [comment]
Only game time (computed by adding globalstep `dtime`s) makes sense, since that is what `minetest.after` uses and we have to be consistent with that, so not the equivalent of `get_us_time`.

This also means results are consistent during a serverstep, otherwise calling a few microseconds later would give you a different time, and I don't think we want that.

If we used `get_us_time`, pausing the game would have no effect. Your weapon would reload while the game is paused.

BTW, if you want to have a go at this yourself, this should be a pretty straightforward addition in `builtin/common/after.lua` (~10 loc maybe?)

**Comment by Zughy** [comment]
@appgurueu thank you for the clarification :) if you concept approve it, I'll try :D

**Comment by sfan5** [comment]
I see two problems with this:
* "elapsed time" is confusing. intuitively I would expect this to be the time the job took to execute
* the usecase seems a bit weird
  * if this is useful: why should it be restricted to queued jobs? wouldn't it be better to expose gametime with the needed precision to allow mods to do this?
  * if it's not useful/too niche: we shouldn't have it in builtin

**Comment by Zughy** [comment]
* `current_time`, `living_time`?
* another usecase: time race, you have 20s to reach the end. An after starts and if you reach the end in e.g. 12.3s, the game will save your record in a leaderboard. If you don't, something will happen when the after ends. I can't think of any usecase outside the after

**Comment by Desour** [comment]
Imo a function for getting the game time (i.e. by exposing the time value used by core.after) would be be better.
It would be more general (e.g. would also work with node timers, and other mods' timers), and a `get_elapsed_time()` would probably be implemented using such a function anyways (<https://github.com/minetest/minetest/pull/15058#discussion_r1732662544>).

**Comment by Desour** [comment]
Oops, I actually wanted to implement this but never came to it. (Adding it to a milestone this time so I don't forget. (Doesn't mean it can't be implemented before.))

---

When thinking about implementing this, I thought, we should also add a function the get the us time of the start of the last server step, in addition to the dtime sum (what after does). I.e.: `core.get_us_time_last_step()` and `core.get_dtime_sum_last_step()`.
But then `core.get_dtime_sum_last_step()` would not always be the same as what `core.after` uses: Not for functions that run in the step but before the after globalstep.
One could maybe use `core.get_dtime_sum_last_step()` as time for `core.after`, but this would change behaviour (i.e. in master, `core.after(0.001,...)` will run in the same server step, with the change not, but it would with given time=0).

So, `core.after` calls before the core.after globalstep can be called in the same step.
(This is not that much of an edge-case. Node timers and ABMs are for example called before globalstep.)
And the actually elapsed time can be shorter than the requested time (well, this is always the case, see also our docs).

Should we "fix" this, by using the actual start time of the step in core.after? Or should we define step start time to be when core.after runs?
(I'm leaving towards the former (also because for example node timer clock for example would be weird with the latter way). But idk if I'm missing something.)

**Comment by sfan5** [comment]
Isn't the solution literally just this? I can't think of a problem with it:
```diff
diff --git a/builtin/common/after.lua b/builtin/common/after.lua
index c116f1200..7e1af1db6 100644
--- a/builtin/common/after.lua
+++ b/builtin/common/after.lua
@@ -114,6 +114,10 @@ end
 local time = 0.0
 local time_next = math.huge
 
+function core.get_global_time()
+       return time
+end
+
 core.register_globalstep(function(dtime)
        time = time + dtime
 
```

**Comment by sfan5** [comment]
I'm not sure I follow. In which concrete situation would this pose a problem?

**Comment by Desour** [comment]
Idk if it will cause problems in practice or not.

What is happening now is that if you use core.after in a nodetimer callback for example, it will run in the same server step, and only a very short time (as measured eg. by `get_us_time`) after (possibly shorter than the given time, but that's always the case).
And if you'd use `core.get_global_time`, it would look like you're still in the server step from before.

Maybe an issue: If you want to use `core.get_global_time` for node timers (the same way as you'd do in core.after, i.e. get the time when you start the timer, and get it again in the callback, and calc the diff), you might get a too small value.

**Comment by sfan5** [comment]
Hmm right. Node timers have a minor accuracy problem anyway:
https://github.com/luanti-org/luanti/blob/fa0c09d202537acbd267a38958f16c883359baca/src/serverenvironment.cpp#L977-L980
But I think we can just document that `get_global_time` only increases once per server step and modders can live wth it.

**Comment by sfan5** [comment]
So should we go ahead with this solution?

**Comment by Desour** [comment]
Yes, I guess. (Document that `get_global_time` and `after` trigger on globalstep or so.)
Would you prefer me or you to open the PR?

**Comment by sfan5** [comment]
Feel free to do it.

**Comment by Desour** [comment]
#16299

**Comment by SmallJoker** [comment]
Coming from #16299

@Zughy is `core.get_server_uptime()` sufficient for your use-case?

**Comment by Zughy** [comment]
Heh, it's been almost a year, I barely remember what I wanted to do, I only remember I had to find a workaround. I can't really answer

**Comment by sfan5** [comment]
I'll remove this from the milestone and set "low priority" since there are easy workarounds.
